packages/salesforcedx-vscode-apex-debugger/src/core/streamingClient.ts

```diff
@@ -0,0 +1,222 @@
+  FileName?: string;
+  Line?: number;
+  Stacktrace?: string;
+}
+
+export interface DebuggerMessage {
+  event: StreamingEvent;
+  sobject: ApexDebuggerEvent;
+}
+
+export class StreamingClientInfo {
+  public channel: string;
+  public timeout: number;
+  public errorHandler: Function;
+  public connectedHandler: Function;
+  public disconnectedHandler: Function;
+  public messageHandler: Function;
+
+  public constructor(builder: StreamingClientInfoBuilder) {
+    this.channel = builder.channel;
+    this.timeout = builder.timeout;
+    this.errorHandler = builder.errorHandler;
+    this.connectedHandler = builder.connectedHandler;
+    this.disconnectedHandler = builder.disconnectedHandler;
+    this.messageHandler = builder.messageHandler;
+  }
+}
+
+export class StreamingClientInfoBuilder {
+  public channel: string;
+  public timeout: number;
+  public errorHandler: Function;
+  public connectedHandler: Function;
+  public disconnectedHandler: Function;
+  public messageHandler: Function;
+
+  public forChannel(channel: string): StreamingClientInfoBuilder {
+    this.channel = channel;
+    return this;
+  }
+
+  public withTimeout(durationInSeconds: number): StreamingClientInfoBuilder {
+    this.timeout = durationInSeconds || DEFAULT_STREAMING_TIMEOUT;
+    return this;
+  }
+
+  public withErrorHandler(handler: Function): StreamingClientInfoBuilder {
+    this.errorHandler = handler;
+    return this;
+  }
+
+  public withConnectedHandler(handler: Function): StreamingClientInfoBuilder {
+    this.connectedHandler = handler;
+    return this;
+  }
+
+  public withDisconnectedHandler(
+    handler: Function
+  ): StreamingClientInfoBuilder {
+    this.disconnectedHandler = handler;
+    return this;
+  }
+
+  public withMsgHandler(handler: Function): StreamingClientInfoBuilder {
+    this.messageHandler = handler;
+    return this;
+  }
+
+  public build(): StreamingClientInfo {
+    return new StreamingClientInfo(this);
+  }
+}
+
+export class StreamingClient {
+  private client: FayeClient;
+  private connected = false;
+  private shouldDisconnect = false;
+  private isReplaySupported = false;
+  private replayId = -1;
+  private clientInfo: StreamingClientInfo;
+
+  public constructor(
+    url: string,
+    accessToken: string,
+    clientInfo: StreamingClientInfo
+  ) {
+    this.clientInfo = clientInfo;
+    this.client = new FayeClient(url, {
+      timeout: this.clientInfo.timeout
+    });
+    this.client.setHeader('Authorization', `OAuth ${accessToken}`);
+    this.client.setHeader('Content-Type', 'application/json');
+  }
+
+  public async subscribe(): Promise<void> {
+    let subscribeAccept: Function, subscribeReject: Function;
+    const returnPromise = new Promise<void>((resolve, reject) => {
+      subscribeAccept = resolve;
+      subscribeReject = reject;
+    });
+
+    this.client.on('transport:down', async () => {
+      if (this.connected) {
+        this.connected = false;
+        this.sendSubscribeRequest();
```

Comment:
transport:down is emitted when a client's request times out, suggesting the server is unreachable, so it's a terminal failure. I've updated the reconnect logic to depend on /meta/connect. Section 8.3.8 of https://docs.cometd.org/current/reference/#_javascript_subscribe.